crates/tracing/src/stack.rs

```diff
@@ -0,0 +1,155 @@
+//! This module is centered around a thread-local Stack of Span references,
+//! with some public functions for pushing and accessing values from the stack.
+//! This is primarily used by the autotrace proc macro in holochain_tracing_macros,
+//! to automatically push new child spans onto the stack when entering a new function frame.
+//! Using functions like `with_top` allow the user to access the stack directly, for situations
+//! like needing to take a span context and send it into another thread, or out of the process entirely.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::Span;
+
+/// This enum defines how to handle situations where we expect there to be a Span
+/// on the stack, but there is none.
+#[allow(dead_code)]
+enum Mode {
+    /// Panic when finding an empty stack. Useful for quickly discovering gaps in tracing coverage
+    Panic,
+    /// Emit a warning and a full backtrace. Note, this is very noisy and slow!
+    Backtrace,
+    /// Ignore cases of an empty stack, and just return a null span.
+    Noop,
+}
+
+const MODE: Mode = Mode::Noop;
+
+thread_local! {
+    static SPANSTACK: RefCell<SpanStack> = RefCell::new(SpanStack::default());
```

Comment:
`RefCell` seems like a reasonable choice here but curious to know why you picked it specifically.

Reply:
@struktured it needs to be able to hand out mutable references arbitrarily, which is what RefCell is for. Basically bypassing the borrow-checker when you can assert that you'll never have problems synchronizing mutable access. Which we can't since this is thread-local. But is there another wrapper that could also work here?

Reply:
>> Which we can't since this is thread-local.
>
>That's exactly it. We technically shouldn't need it.

The compiler needs it though. `static`s are immutable, so you can only get immutable refs to them.

Reply:
> The compiler needs it though. statics are immutable, so you can only get immutable refs to them. :+1: